nav2_bt_navigator/src/bt_navigator.cpp

```diff
@@ -241,109 +170,74 @@ nav2_util::CallbackReturn
-      feedback_msg->number_of_recoveries = recovery_count;
-      feedback_msg->navigation_time = now() - start_time_;
-      action_server_->publish_feedback(feedback_msg);
-    };
-
-  // Execute the BT that was previously created in the configure step
-  nav2_behavior_tree::BtStatus rc = bt_->run(&tree_, on_loop, is_canceling);
-  // Make sure that the Bt is not in a running state from a previous execution
-  // note: if all the ControlNodes are implemented correctly, this is not needed.
-  bt_->haltAllActions(tree_.rootNode());
-
-  switch (rc) {
-    case nav2_behavior_tree::BtStatus::SUCCEEDED:
-      RCLCPP_INFO(get_logger(), "Navigation succeeded");
-      action_server_->succeeded_current();
-      break;
-
-    case nav2_behavior_tree::BtStatus::FAILED:
-      RCLCPP_ERROR(get_logger(), "Navigation failed");
-      action_server_->terminate_current();
-      break;
-
-    case nav2_behavior_tree::BtStatus::CANCELED:
-      RCLCPP_INFO(get_logger(), "Navigation canceled");
-      action_server_->terminate_all();
-      break;
-  }
+  auto blackboard = bt_action_server_->getBlackboard();
+  blackboard->set<int>("number_recoveries", 0);  // NOLINT
+  // Update the goal pose on the blackboard
+  blackboard->set<geometry_msgs::msg::PoseStamped>("goal", goal->pose);
+
+  feedback_msg_ = std::make_shared<Action::Feedback>();
+
+  return true;
 }
 
 void
-BtNavigator::initializeGoalPose()
+BtNavigator::onLoop()
 {
-  auto goal = action_server_->get_current_goal();
+  // action server feedback (pose, duration of task,
+  // number of recoveries, and distance remaining to goal)
+  nav2_util::getCurrentPose(
+    feedback_msg_->current_pose, *tf_, global_frame_, robot_frame_, transform_tolerance_);
 
-  RCLCPP_INFO(
-    get_logger(), "Begin navigating from current location to (%.2f, %.2f)",
-    goal->pose.pose.position.x, goal->pose.pose.position.y);
+  auto blackboard = bt_action_server_->getBlackboard();
 
-  // Reset state for new action feedback
-  start_time_ = now();
-  blackboard_->set<int>("number_recoveries", 0);  // NOLINT
+  geometry_msgs::msg::PoseStamped goal_pose;
+  blackboard->get("goal", goal_pose);
 
-  // Update the goal pose on the blackboard
-  blackboard_->set<geometry_msgs::msg::PoseStamped>("goal", goal->pose);
+  feedback_msg_->distance_remaining = nav2_util::geometry_utils::euclidean_distance(
+    feedback_msg_->current_pose.pose, goal_pose.pose);
+
+  int recovery_count = 0;
+  blackboard->get<int>("number_recoveries", recovery_count);
+  feedback_msg_->number_of_recoveries = recovery_count;
+  feedback_msg_->navigation_time = now() - start_time_;
+
+  bt_action_server_->getActionServer()->publish_feedback(feedback_msg_);
+}
+
+void
+BtNavigator::onPreempt()
+{
+  RCLCPP_INFO(get_logger(), "Received goal preemption request");
+  auto action_server = bt_action_server_->getActionServer();
+  action_server->accept_pending_goal();
+  onGoalReceived(action_server->get_current_goal());
```

Comment:
This function also calls the BT loading step which shouldn't be necessarily called in preemption cases. That's a topic of contention right now that I don't want addressed in this PR doing alot of other updates.

Reply:
This was part of the reason I wanted a separate preemption function, so we could treat them differently

Reply:
Yup, got it. Should be okay now.